pkg/process/metadata/parser/dockerproxy.go

```diff
@@ -28,55 +28,61 @@ type proxy struct {
 	target model.ContainerAddr
 }
 
-// NewFilter instantiates a new filter loaded with docker-proxy instance information
-func NewFilter(procs map[int32]*procutil.Process) *Filter {
-	filter := new(Filter)
-	filter.LoadProxies(procs)
-	return filter
+// NewDockerProxy instantiates a new filter loaded with docker-proxy instance information
+func NewDockerProxy() *DockerProxy {
+	return &DockerProxy{
+		proxyByPID:    make(map[int32]*proxy),
+		proxyByTarget: make(map[model.ContainerAddr]*proxy),
+	}
 }
 
-// LoadProxies by inspecting processes information
-func (f *Filter) LoadProxies(procs map[int32]*procutil.Process) {
-	f.proxyByPID = make(map[int32]*proxy)
-	f.proxyByTarget = make(map[model.ContainerAddr]*proxy)
+func (d *DockerProxy) Extract(processes map[int32]*procutil.Process) {
+	proxyByPID := make(map[int32]*proxy)
+	proxyByTarget := make(map[model.ContainerAddr]*proxy)
 
-	for _, p := range procs {
-		proxy := extractProxyTarget(p)
-		if proxy == nil {
+	for _, p := range processes {
+		if proxy, seen := d.proxyByPID[p.Pid]; seen {
+			proxyByPID[p.Pid] = proxy
+			proxyByTarget[proxy.target] = proxy
 			continue
 		}
 
-		log.Debugf("detected docker-proxy with pid=%d target.ip=%s target.port=%d target.proto=%s",
-			proxy.pid,
-			proxy.target.Ip,
-			proxy.target.Port,
-			proxy.target.Protocol,
-		)
-
-		// Add proxy to cache
-		f.proxyByPID[proxy.pid] = proxy
-		f.proxyByTarget[proxy.target] = proxy
+		if proxy := extractProxyTarget(p); proxy != nil {
+			log.Debugf("detected docker-proxy with pid=%d target.ip=%s target.port=%d target.proto=%s",
+				proxy.pid,
+				proxy.target.Ip,
+				proxy.target.Port,
+				proxy.target.Protocol,
+			)
+
+			// Add proxy to cache
+			proxyByPID[p.Pid] = proxy
+			proxyByTarget[proxy.target] = proxy
+		}
 	}
+
+	d.proxyByPID = proxyByPID
+	d.proxyByTarget = proxyByTarget
 }
 
 // Filter all connections that have a docker-proxy at one end
-func (f *Filter) Filter(payload *model.Connections) {
-	if len(f.proxyByPID) == 0 {
+func (d *DockerProxy) Filter(payload *model.Connections) {
```

Comment:
Correct me if I'm wrong but `Filter` and `Extract` should be executed by the same thread otherwise there will be a race condition on `proxyByPID` and `proxyByTarget`, right?

It's not a bit deal now since they both run in the same check run. I wonder if we should let it documented though to prevent accidental usages of it. Have you decided to not introduce a mutex here due to performance considerations?

Reply:
You are right, but we can leave this as such until we improve upon this in the future.